core/src/main/scala/kafka/controller/PartitionStateMachine.scala

```diff
@@ -102,6 +105,18 @@ class PartitionStateMachine(controller: KafkaController) extends Logging {
     info("Stopped partition state machine")
   }
 
+  def executeFutureCallbacks(listOfFutures: Iterable[Future[() => Unit]]): Unit = {
+    import scala.concurrent.duration._
+    val futureOfList = Future.sequence(listOfFutures)
```

Comment:
`Futures.sequence` [folds](https://github.com/scala/scala/blob/2.11.x/src/library/scala/concurrent/Future.scala#L499) over a provided list of futures chaining them via `flatMap` which doesn't allow GC to collect them individually after completion. Here's a simple implementation that doesn't has this problem: [gist](https://gist.github.com/ataraxer/7d49e49e90ba025dbcdf679c39297741).